### ISSUE TITLE
Support articulations

### ISSUE THREAD

**Issue by joeberkovitz** [w3c/mnx#103]
This issue notes the need for articulations, including accents, staccato, tenuto and many others supported by MusicXML.

**Comment by adrianholovaty** [comment]
I'd like to get articulations designed as the next big MNX improvement, especially since the data model is relatively simple. Some thoughts on various details:

**Place in the document.** An articulation should live on an event, not on a note. This seems to be the semantically correct place — it's impossible (or at least really weird) for an articulation to be active on a single note of a chord. I discussed this with @dspreadbury and @mscuthbert at the last co-chair meeting and they said this was the right approach.

**Metadata.** We should allow for metadata to be attached to each articulation. I was originally tempted to use simple booleans on an event object, such as `"staccato": true`, but that would mean we couldn't control an articulation's position, color, etc. Best to use a whole object for the articulation, following the general convention that "each distinct notational object gets a JSON object."

**Allowing (or disallowing) weird combinations.** Technically some articulations are mutually exclusive — like "accent" and "strong-accent" from MusicXML — but I don't think we need to bend over backwards to enforce this in the schema. We could simply use documentation to say something like "If an event has both 'accent' and 'strong-accent', prefer 'strong-accent'".

**Which articulations to include.** I think it makes sense to copy MusicXML's list — [see here](https://w3c.github.io/musicxml/musicxml-reference/elements/articulations/) — except for `<other-articulation>`. That one is a bit too open-ended and not semantic enough. If people use that one in practice, I'm curious to hear arguments for it! But the burden should be on them to make the case for it.

**Grouping articulations in a container object.** Should articulations live within a "container" object within the event, or should they live directly on the event?

With a container:
```json
{
  "type": "event",
  "notes": [...],
  "articulations": [
    {"type": "staccato"},
    {"type": "tenuto"}
  ]
}
```

Without a container:
```json
{
  "type": "event",
  "notes": [...],
  "staccato": {},
  "tenuto": {}
}
```

A pro of the container: It could provide a single place for metadata about _all_ the event's articulations. For example, color or position. With that said, we'd still need the ability to set color or position on an _individual_ articulation — it's possible for one articulation to be above the staff while another is below the staff (for the same event). So maybe higher-level metadata is not actually useful in practice?

Another pro of the container: The hierarchy means event objects would perhaps feel better organized...? This is more of a matter of personal taste.

A con of the container: Nothing prevents people from putting two `"type": "staccato"`s within a single `"articulations"` list. The container-less approach solves that structurally.

Another con of the container: It's more verbose.

**Final thoughts.** Having pondered this for a few weeks, I'd suggest going with the containerless approach, like this:

```json
{
  "type": "event",
  "notes": [...],
  "staccato": {},
  "tenuto": {}
}
```

Nice and light, while still allowing for per-articulation metadata.

Thoughts welcome!

**Comment by williamclocksin** [comment]
Some thoughts.  One question is whether per-articulation metadata knows the context of the other articulations associated with the event.  If it doesn’t, then maybe the container is needed so that software can place the articulations according to context. In Calliope25, multiple articulation symbols are in a stack that is associated with an event.  The stack can be placed above or below the staff, or at the head or tail of a note, or in combinations. Articulations within a stack can be re-ordered.  Stacks can also be laid out sideways, but I’ve never needed to use that. Events can also have multiple articulations associated with them.  This gives a reasonably flexible provision, but I’m not sure how to represent all this in MNX, containers or otherwise. Using MNX for data serialisation means that we need to ensure that X equals read(write(X)).
Also, I don’t think it is necessary to enforce musical meanings in the representation. The composer should be free to specify a combination of articulations that we think are weird but maybe is meaningful to the composer.  Some music audio performance devices try to render articulations, but it is an approximation at best, and we can’t predict what weirdness might be useful to the composer/user.

**Comment by adrianholovaty** [comment]
Just pointing out another option from @mscuthbert — using an object, instead of a list, for a container:

```json
{
  "type": "event",
  "notes": [...],
  "articulations": {
    "staccato": {},
    "tenuto": {}
  }
}
```

And another pro for the nested approach is it keeps the top-level event from feeling overly big and complicated.

**Comment by mscuthbert** [comment]
One possibility that Daniel, Adrian, and I discussed today (that I'm leaning in favor of, but definitely no true consensus reached), is that an Event structures its dependent "articulations" not so much on the basis of whether we classify them as Articulations or not (or as playing indications, etc.) but by where they are engraved and what engraving rules apply to them.  So there's a conceptual space on the notehead side (and a less important but still crucial one on the stem side) where Articulations are placed (staccato, accent, etc.) but this same vertical space (with different default positioning and ordering) is also occupied by other things like bow marks, pizz marks, etc.  Whereas the musicxml articulations of breath mark, caesura, falloff, doit, etc. occupy different spacing positions.  Breath mark and caesura: before the next note.  Falloff, doit: next to the notehead [but which one in the unusual case that there's a chord?].  

Flutter tonguing is definitely an articulation in the conventional sense, but it is not notated by a mark in the "articulation" space.  It's either with a tremolo or "Z" mark on the stem, or a flzg. mark in the "expressions" space, possibly with a line after it (like a spanner) in the 8va space.  So for this reason it's never been included in musicxml articulations.  But any playback engine should definitely consider it an articulation.

So perhaps we want to separate entry-attached elements by location?  While keeping the semantic meaning of "staccato" "accent" etc.  So the "articulations" element might become "articulations_location" or simply call it "articulations"? or abbreviate slightly as "articulations_loc"

**Comment by mscuthbert** [comment]
Consensus is definitely "Caesura" is not an articulation.  Breath mark that affects the player only might be an articulation (or might not).  Breath mark that affects overall tempo, however, is definitely not an articulation.

**Comment by adrianholovaty** [comment]
I've just opened pull request #322, with basic support for event markings — including articulations.

**Comment by adrianholovaty** [comment]
I've just merged #322, which adds support for event markings, including articulations. The commit is f9ab65464, and here's the new docs page for markings: https://w3c.github.io/mnx/docs/mnx-reference/objects/event-markings/
